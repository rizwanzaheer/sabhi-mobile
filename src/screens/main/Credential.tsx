import React, { useEffect, useState } from 'react'
import {
  Container,
  Credential,
  Device,
  Text,
  RadioBtn,
  Constants,
  Typings,
} from '@kancha/kancha-ui'
import { NavigationStackScreenProps } from 'react-navigation-stack'
import { HeaderButtons, Item } from 'react-navigation-header-buttons'
import { FlatList } from 'react-native'
import { Colors } from '../../theme'
import { ScrollView } from 'react-native-gesture-handler'

interface CredentialStyle {
  background: Typings.BrandPropOptions
  shadow: number
}

interface Props extends NavigationStackScreenProps {}

const CredentialDetail: React.FC<Props> & { sharedElements: any } & {
  navigationOptions: any
} = ({ navigation }) => {
  const credentials = navigation.getParam('credentials')
  const initialCredential = navigation.getParam('credentialIndex', 0)
  const [sharingMode, toggleSharingMode] = useState<boolean>(false)
  const [selected, updateSelected] = useState<number[]>([])

  const isSelected = (index: number) => {
    return selected.includes(index)
  }

  const selectedStyle = (index: number): CredentialStyle => {
    return {
      background: isSelected(index) ? 'primary' : 'primary',
      shadow: 0,
    }
  }

  const selectCredential = (index: number) => {
    if (selected.includes(index)) {
      updateSelected(selected.filter((item: number) => item !== index))
    } else {
      updateSelected(selected.concat([index]))
    }
  }

  const initailOffset = { x: (Device.width - 15) * initialCredential, y: 0 }

  useEffect(() => {
    navigation.setParams({ sharingMode, toggleSharingMode })
  }, [sharingMode])

  const renderCredential = ({
    item,
    index,
  }: {
    item: Typings.VerifiableCredential & {
      raw: string
      issuer: any
      subject: any
      claims: any[]
      expirationDate: any
    }
    index: number
  }) => (
    <Container w={Device.width - 10} padding paddingRight={10}>
      <ScrollView testID={'SCROLLVIEW'} scrollEventThrottle={16}>
        <Credential
          testID={'CREDENTIAL'}
          onPress={() => sharingMode && selectCredential(index)}
          detailMode
          jwt={item.raw}
          issuer={item.issuer}
          subject={item.subject}
          fields={item.claims}
          exp={item.expirationDate}
          {...selectedStyle(index)}
        />
        {sharingMode && (
          <RadioBtn
            testID={'RADIO_BTN'}
            selected={isSelected(index)}
            onPress={() => selectCredential(index)}
          >
            {isSelected(index) && 'Sharing'}
          </RadioBtn>
        )}
      </ScrollView>
    </Container>
  )

  return (
    <Container flex={1} backgroundColor={Colors.BLACK}>
      {sharingMode && (
        <Container padding>
          <Text type={Constants.TextTypes.Body} textColor={Colors.LIGHT_GREY}>
            Select credential(s) for sharing. Flow to be completed.
          </Text>
        </Container>
      )}
      <FlatList
        testID={'FLATLIST'}
        horizontal
        pagingEnabled
        contentOffset={initailOffset}
        showsHorizontalScrollIndicator={false}
        snapToAlignment={'center'}
        keyExtractor={(item: Typings.VerifiableCredential) => item.hash}
        data={credentials}
        renderItem={renderCredential}
      ></FlatList>
    </Container>
  )
}

CredentialDetail.navigationOptions = ({ navigation }: any) => {
  const { sharingMode, toggleSharingMode } = navigation.state.params || {}
  const sharingModeEnabled = navigation.getParam('sharingModeEnabled', true)

  return {
    title: 'Credential',
    headerStyle: {
      backgroundColor: '#f4511e',
    },
    headerLeft: () => (
      <HeaderButtons>
        <Item
          testID={'DONE_BTN'}
          title={'Done'}
          onPress={navigation.dismiss}
          color={Colors.WHITE}
        />
      </HeaderButtons>
    ),
    headerRight: () => (
      <HeaderButtons>
        {sharingMode && sharingModeEnabled ? (
          <Item
            testID={'CANCEL_BTN'}
            title={'Cancel'}
            onPress={() => toggleSharingMode(false)}
            color={Colors.WHITE}
          />
        ) : sharingModeEnabled ? (
          <Item
            testID={'SHARE_BTN'}
            title={'Share'}
            onPress={() => toggleSharingMode(true)}
            color={Colors.WHITE}
          />
        ) : null}
      </HeaderButtons>
    ),
  }
}

CredentialDetail.sharedElements = (navigation: any) => {
  return navigation
    .getParam('credentials')
    .map((vc: Typings.VerifiableCredential) => ({
      id: vc.hash,
      animation: 'fade',
      resize: 'clip',
    }))
}

export default CredentialDetail
