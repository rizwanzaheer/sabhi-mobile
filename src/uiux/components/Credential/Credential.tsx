import * as React from 'react'
import Container from '../Container/Container'
import Text, { TextTypes } from '../Text/Text'
import Avatar from '../Avatar/Avatar'
import Card from '../Card/Card'
import Icon from '../Icon/Icon'
import * as Kancha from '../../types'
import { withTheme } from '../../theming'
import QRCode from 'react-native-qrcode-svg'

const S = require('sugar/string')

interface Field {
  type: string
  value: any
  isObj: boolean
}

export interface CredentialProps {
  onPress?: () => void
  issuer: Kancha.Identity
  subject: Kancha.Identity
  exp: number
  jwt?: string
  qrSize?: number
  fields: Field[]
  testID?: string
  shadow?: number
  background?: Kancha.BrandPropOptions
  detailMode?: boolean
  marginBottom?: number
  theme: any
}

const Credential: React.FC<CredentialProps> = ({
  shadow,
  onPress,
  issuer,
  background,
  fields,
  jwt,
  qrSize,
  subject,
  testID,
  detailMode,
  theme,
  marginBottom,
}) => {
  const subProfileSource = subject.profileImage ? { source: { uri: subject.profileImage } } : {}
  const issProfileSource = issuer.profileImage ? { source: { uri: issuer.profileImage } } : {}
  const credentialFields = fields.filter((field: any, i) => (detailMode ? field : i < 2))

  return (
    <Card
      onPress={onPress}
      testID={testID}
      shadow={shadow || 0}
      background={background}
      marginBottom={marginBottom}
    >
      <Container flexDirection={'row'} alignItems={'flex-start'}>
        <Container>
          <Avatar
            border
            {...subProfileSource}
            address={subject.did}
            type={'circle'}
            gravatarType={'retro'}
            size={40}
          />
          <Container viewStyle={{ position: 'absolute', top: 0, left: 15 }}>
            <Avatar
              border
              {...issProfileSource}
              address={issuer.did}
              type={'circle'}
              gravatarType={'retro'}
              size={40}
            />
          </Container>
        </Container>
        <Container marginLeft={28}>
          <Container>
            <Text bold type={TextTypes.ActivityTitle}>
              {issuer.shortId}
            </Text>
            <Container marginTop={3} flexDirection={'row'} alignItems={'flex-start'}>
              <Container marginRight={5}>
                <Icon
                  icon={{ name: 'ios-play', iconFamily: 'Ionicons' }}
                  size={15}
                  color={theme.colors.primary.text}
                />
              </Container>
              <Text type={TextTypes.ActivityTitle} textStyle={{ fontSize: 15 }}>
                {subject.shortId}
              </Text>
            </Container>
          </Container>
        </Container>
      </Container>
      <Container marginTop={detailMode ? 30 : 16}>
        {credentialFields.map((field: any, i: number) => {
          const fieldValueImage = !field.isObj
            ? field.value.endsWith('.jpg') || field.value.endsWith('.png')
            : false
          return (
            <Container marginBottom={5} key={i}>
              <Container>
                <Text textStyle={{ fontSize: 12, textTransform: 'uppercase' }} type={TextTypes.SubTitle}>
                  {S.String.titleize(field.type)}
                </Text>
              </Container>
              <Container justifyContent={'flex-start'}>
                {fieldValueImage ? (
                  <Container paddingTop={5}>
                    <Avatar
                      source={{ uri: field.value }}
                      address={issuer.did}
                      type={'rounded'}
                      gravatarType={'retro'}
                      size={25}
                    />
                  </Container>
                ) : (
                  <Text type={TextTypes.ActivityTitle}>
                    {field.isObj ? 'Type not supported yet' : field.value}
                  </Text>
                )}
              </Container>
            </Container>
          )
        })}
        {!detailMode && fields.length > 2 && (
          <Container>
            <Text type={TextTypes.SubTitle}>...</Text>
          </Container>
        )}
        {detailMode && jwt && (
          <Container marginTop={50} alignItems={'center'} justifyContent={'center'}>
            <QRCode size={qrSize ? qrSize : 200} value={jwt} />
          </Container>
        )}
      </Container>
    </Card>
  )
}

export default withTheme(Credential)
