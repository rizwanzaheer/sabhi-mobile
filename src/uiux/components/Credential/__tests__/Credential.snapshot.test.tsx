import * as React from 'react'
import { render, fireEvent } from 'react-native-testing-library'
import Credential from '../Credential'

const issuer = { did: '0xf1234', profileImage: 'http://', shortId: 'Issuer Name' }
const subject = { did: '0xf1234', profileImage: 'http://', shortId: 'Issuer Name' }
const fields = [
  {
    type: 'name',
    value: 'test value',
    isObj: false,
  },
]

describe('Component(snapshot): Credential', () => {
  it('should render with default props', () => {
    const tree = render(
      <Credential issuer={issuer} exp={1575299770} fields={fields} subject={subject} />,
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should fire on press event', () => {
    const onPress = jest.fn()
    const { getByTestId } = render(
      <Credential
        marginBottom={10}
        issuer={issuer}
        exp={1575299770}
        fields={fields}
        subject={subject}
        onPress={onPress}
        testID={'CREDENTIAL_BTN'}
      />,
    )

    fireEvent.press(getByTestId('CREDENTIAL_BTN'))
    expect(onPress).toHaveBeenCalled()
  })
})
