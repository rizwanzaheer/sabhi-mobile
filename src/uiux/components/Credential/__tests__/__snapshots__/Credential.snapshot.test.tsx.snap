// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component(snapshot): Credential should render with default props 1`] = `
<View
  accessible={true}
  focusable={false}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": undefined,
      "borderRadius": 5,
      "marginBottom": 15,
    }
  }
>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": undefined,
        "borderRadius": undefined,
        "flex": undefined,
        "flexDirection": undefined,
        "height": undefined,
        "justifyContent": undefined,
        "margin": undefined,
        "marginBottom": 15,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": 15,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  >
    <View
      pointerEvents="auto"
      style={
        Object {
          "alignItems": "flex-start",
          "backgroundColor": undefined,
          "borderRadius": undefined,
          "flex": undefined,
          "flexDirection": "row",
          "height": undefined,
          "justifyContent": undefined,
          "margin": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "width": undefined,
        }
      }
    >
      <View
        pointerEvents="auto"
        style={
          Object {
            "alignItems": undefined,
            "backgroundColor": undefined,
            "borderRadius": undefined,
            "flex": undefined,
            "flexDirection": undefined,
            "height": undefined,
            "justifyContent": undefined,
            "margin": undefined,
            "marginBottom": undefined,
            "marginLeft": undefined,
            "marginRight": undefined,
            "marginTop": undefined,
            "padding": undefined,
            "paddingBottom": undefined,
            "paddingHorizontal": undefined,
            "paddingLeft": undefined,
            "paddingRight": undefined,
            "paddingTop": undefined,
            "width": undefined,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "http://",
            }
          }
          style={
            Array [
              Object {
                "backgroundColor": "#0078ff",
                "borderColor": "#FFFFFF",
                "borderRadius": 20,
                "borderWidth": 2,
                "height": 40,
                "width": 40,
              },
              Object {
                "backgroundColor": "transparent",
              },
            ]
          }
        />
        <View
          pointerEvents="auto"
          style={
            Object {
              "alignItems": undefined,
              "backgroundColor": undefined,
              "borderRadius": undefined,
              "flex": undefined,
              "flexDirection": undefined,
              "height": undefined,
              "justifyContent": undefined,
              "left": 15,
              "margin": undefined,
              "marginBottom": undefined,
              "marginLeft": undefined,
              "marginRight": undefined,
              "marginTop": undefined,
              "padding": undefined,
              "paddingBottom": undefined,
              "paddingHorizontal": undefined,
              "paddingLeft": undefined,
              "paddingRight": undefined,
              "paddingTop": undefined,
              "position": "absolute",
              "top": 0,
              "width": undefined,
            }
          }
        >
          <Image
            source={
              Object {
                "uri": "http://",
              }
            }
            style={
              Array [
                Object {
                  "backgroundColor": "#0078ff",
                  "borderColor": "#FFFFFF",
                  "borderRadius": 20,
                  "borderWidth": 2,
                  "height": 40,
                  "width": 40,
                },
                Object {
                  "backgroundColor": "transparent",
                },
              ]
            }
          />
        </View>
      </View>
      <View
        pointerEvents="auto"
        style={
          Object {
            "alignItems": undefined,
            "backgroundColor": undefined,
            "borderRadius": undefined,
            "flex": undefined,
            "flexDirection": undefined,
            "height": undefined,
            "justifyContent": undefined,
            "margin": undefined,
            "marginBottom": undefined,
            "marginLeft": 28,
            "marginRight": undefined,
            "marginTop": undefined,
            "padding": undefined,
            "paddingBottom": undefined,
            "paddingHorizontal": undefined,
            "paddingLeft": undefined,
            "paddingRight": undefined,
            "paddingTop": undefined,
            "width": undefined,
          }
        }
      >
        <View
          pointerEvents="auto"
          style={
            Object {
              "alignItems": undefined,
              "backgroundColor": undefined,
              "borderRadius": undefined,
              "flex": undefined,
              "flexDirection": undefined,
              "height": undefined,
              "justifyContent": undefined,
              "margin": undefined,
              "marginBottom": undefined,
              "marginLeft": undefined,
              "marginRight": undefined,
              "marginTop": undefined,
              "padding": undefined,
              "paddingBottom": undefined,
              "paddingHorizontal": undefined,
              "paddingLeft": undefined,
              "paddingRight": undefined,
              "paddingTop": undefined,
              "width": undefined,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#333333",
                "flexWrap": "wrap",
                "fontSize": 17,
                "fontWeight": "bold",
              }
            }
          >
            Issuer Name
          </Text>
          <View
            pointerEvents="auto"
            style={
              Object {
                "alignItems": "flex-start",
                "backgroundColor": undefined,
                "borderRadius": undefined,
                "flex": undefined,
                "flexDirection": "row",
                "height": undefined,
                "justifyContent": undefined,
                "margin": undefined,
                "marginBottom": undefined,
                "marginLeft": undefined,
                "marginRight": undefined,
                "marginTop": 3,
                "padding": undefined,
                "paddingBottom": undefined,
                "paddingHorizontal": undefined,
                "paddingLeft": undefined,
                "paddingRight": undefined,
                "paddingTop": undefined,
                "width": undefined,
              }
            }
          >
            <View
              pointerEvents="auto"
              style={
                Object {
                  "alignItems": undefined,
                  "backgroundColor": undefined,
                  "borderRadius": undefined,
                  "flex": undefined,
                  "flexDirection": undefined,
                  "height": undefined,
                  "justifyContent": undefined,
                  "margin": undefined,
                  "marginBottom": undefined,
                  "marginLeft": undefined,
                  "marginRight": 5,
                  "marginTop": undefined,
                  "padding": undefined,
                  "paddingBottom": undefined,
                  "paddingHorizontal": undefined,
                  "paddingLeft": undefined,
                  "paddingRight": undefined,
                  "paddingTop": undefined,
                  "width": undefined,
                }
              }
            >
              <Icon
                color="#333333"
                name="ios-play"
                size={15}
              />
            </View>
            <Text
              style={
                Object {
                  "color": "#333333",
                  "flexWrap": "wrap",
                  "fontSize": 15,
                }
              }
            >
              Issuer Name
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      pointerEvents="auto"
      style={
        Object {
          "alignItems": undefined,
          "backgroundColor": undefined,
          "borderRadius": undefined,
          "flex": undefined,
          "flexDirection": undefined,
          "height": undefined,
          "justifyContent": undefined,
          "margin": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": 16,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "width": undefined,
        }
      }
    >
      <View
        pointerEvents="auto"
        style={
          Object {
            "alignItems": undefined,
            "backgroundColor": undefined,
            "borderRadius": undefined,
            "flex": undefined,
            "flexDirection": undefined,
            "height": undefined,
            "justifyContent": undefined,
            "margin": undefined,
            "marginBottom": 5,
            "marginLeft": undefined,
            "marginRight": undefined,
            "marginTop": undefined,
            "padding": undefined,
            "paddingBottom": undefined,
            "paddingHorizontal": undefined,
            "paddingLeft": undefined,
            "paddingRight": undefined,
            "paddingTop": undefined,
            "width": undefined,
          }
        }
      >
        <View
          pointerEvents="auto"
          style={
            Object {
              "alignItems": undefined,
              "backgroundColor": undefined,
              "borderRadius": undefined,
              "flex": undefined,
              "flexDirection": undefined,
              "height": undefined,
              "justifyContent": undefined,
              "margin": undefined,
              "marginBottom": undefined,
              "marginLeft": undefined,
              "marginRight": undefined,
              "marginTop": undefined,
              "padding": undefined,
              "paddingBottom": undefined,
              "paddingHorizontal": undefined,
              "paddingLeft": undefined,
              "paddingRight": undefined,
              "paddingTop": undefined,
              "width": undefined,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#BBBBBB",
                "flexWrap": "wrap",
                "fontSize": 12,
                "textTransform": "uppercase",
              }
            }
          >
            Name
          </Text>
        </View>
        <View
          pointerEvents="auto"
          style={
            Object {
              "alignItems": undefined,
              "backgroundColor": undefined,
              "borderRadius": undefined,
              "flex": undefined,
              "flexDirection": undefined,
              "height": undefined,
              "justifyContent": "flex-start",
              "margin": undefined,
              "marginBottom": undefined,
              "marginLeft": undefined,
              "marginRight": undefined,
              "marginTop": undefined,
              "padding": undefined,
              "paddingBottom": undefined,
              "paddingHorizontal": undefined,
              "paddingLeft": undefined,
              "paddingRight": undefined,
              "paddingTop": undefined,
              "width": undefined,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#333333",
                "flexWrap": "wrap",
                "fontSize": 17,
              }
            }
          >
            test value
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
